java/src/main/native/src/RmmJni.cpp

```diff
@@ -26,12 +26,120 @@ using rmm::mr::device_memory_resource;
 
 namespace {
 
+// Alignment to which the RMM memory resource will round allocation sizes
+constexpr std::size_t RMM_ALLOC_SIZE_ALIGNMENT = 512;
+
 constexpr char const* RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
 
+/**
+ * @brief An RMM device memory resource that delegates to another resource
+ * while tracking the amount of memory allocated.
+ *
+ * @tparam Upstream Type of memory resource that will be wrapped.
+ * @tparam size_align The size to which all allocation requests are
+ * aligned. Must be a value >= 1.
+ */
+template <typename Upstream>
+class tracking_resource_adaptor final : public device_memory_resource {
+public:
+  /**
+   * @brief Constructs a new tracking resource adaptor that delegates to
+   * `mr` for all allocation operations while tracking the amount of memory
+   * allocated.
+   *
+   * @param mr The resource to use for memory allocation operations.
+   * @param size_alignment The alignment to which the `mr` resource will
+   * round up all memory allocation size requests.
+   */
+  tracking_resource_adaptor(Upstream* mr, std::size_t size_alignment)
+      : resource{mr}, size_align{size_alignment} {}
+
+  Upstream* get_wrapped_resource() {
+    return resource;
+  }
+
+  std::size_t get_total_allocated() {
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+    return total_allocated;
+  }
+
+private:
+  Upstream* const resource;
+  std::size_t const size_align;
+  std::size_t total_allocated{0};
+
+  // map and associated lock to track memory sizes by address
+  // TODO: This should be removed when rmm::alloc and rmm::free are removed and the size parameter
```

Comment:
I think this is true in libcudf now.

Reply:
Unfortunately it's not true everywhere that may call RMM, such as parts in the JNI bindings (or other app code that may call RMM directly).  The JNI layer's RMM usage is being cleaned up as we speak.  However until RMM prevents the possibility of size ever being bogus, which means https://github.com/rapidsai/rmm/issues/302 is finally resolved, I'd like to leave the size-tracking in place to have all the bases covered.

Reply:
I just put up a PR to fix it https://github.com/rapidsai/cudf/pull/4816 because it is now true.